helm-v3/deisgn-proposal.md

```diff
@@ -0,0 +1,715 @@
+# Helm 3 Design Proposal
+
+This document contains a proposal for Helm 3. It assumes deep familiarity with Helm 2, and assumes that much of Helm 2's architecture will simply carry over without major change.
+
```

Comment:
General comment I'm going to put right at the top. after reading this through. The proposal is quite the opposite of the suggestion one of the k8s devs made at the helm summit. Make things more modular and follow the unix/k8s philosophy of having one tool do one thing and do it well.

One of the things I like about helm2 today is its fairly focused. it does basically two things. packaging multiple objects together, and templating those objects.

His idea was to break it up into at least 2 pieces. one that did the packaging of multiple objects, and another that did the templating. This would allow alternate template engines to be easily added for example.

With CRD's, this is totally possible.
upload a package release CRD. it takes it, pulls tarball, and instantiates each k8s object inside. those k8s objects, could be CRD's for a templater, that takes those and values, and templates them out to the real objects.

This would even let you instantiate a templated helm release crd that then instantiated multiple releases. lots of things that are difficult today like templating out values would naturally fall out of that natural breaking up of functionality.

Should we explore that idea further? The discussion at the helm summit seemed to be positive amongst the discussion members.

Reply:
Multiple template engines will be supported via Lua.

The main idea, which is very understated in the proposal (by design) is that we're trying to focus on the chart -> manifest process, and back off of the release management part. There are several tools in development that will take a CRD that defines an application and then install and manage that application (the Application WG is working on these). Our intention is to be able to describe charts that can be used to produce those CRDs, and then let the backing systems manage the ins and outs of actual application lifecycle management.

At this point, we still need to manage a release as a thing. So we can't completely strip out the install/upgrade/rollback/delete concepts. But now we're heading back toward the simplistic notions of those that are part of `kubectl`. So we're de-emphasizing lifecycle management.

Instead, we're now focusing on the workflow that is "load chart, accumulate values, render chart into manifest, (optionally) send manifest". For that reason, the vast majority of the changes in this proposal are all centered on how we make that a flexible, pluggable system.

Finally, keep in mind that this is just one proposal. It's not set in stone, though currently it's the only one that has gotten any agreement.